tests/e2e/tidbcluster/tidbcluster.go

```diff
@@ -1558,8 +1559,313 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
+			framework.ExpectNoError(err, "failed to get ConfigMap %s/%s", ns, tidbCmName)
+			gomega.Expect(tidbCm.Data["config-file"]).To(gomega.ContainSubstring("level = \"info\""))
+		})
+
+		// this case merge scale-in/scale-out into one case, may seems a little bit dense
+		// when scale-in, replica is first set to 5 and changed to 3
+		// when scale-out, replica is first set to 3 and changed to 5
+		ginkgo.Context("while concurrently scale PD", func() {
+			operation := []string{"in", "out"}
+			for _, op := range operation {
+				ginkgo.It(op, func() {
+					ginkgo.By("Deploy initial tc")
+					tcName := fmt.Sprintf("scale-%s-pd-concurrently", op)
+					tc := fixture.GetTidbCluster(ns, tcName, utilimage.TiDBV4Prev)
+					if op == "in" {
+						tc.Spec.PD.Replicas = 5
+					} else {
+						tc.Spec.PD.Replicas = 3
+					}
+					utiltc.MustCreateTCWithComponentsReady(genericCli, oa, tc, 5*time.Minute, 10*time.Second)
+
+					ginkgo.By("Upgrade PD version")
+					err := controller.GuaranteedUpdate(genericCli, tc, func() error {
+						tc.Spec.PD.Version = pointer.StringPtr(utilimage.TiDBV4)
+						return nil
+					})
+					framework.ExpectNoError(err, "failed to update PD version to %q", utilimage.TiDBV4)
+
+					ginkgo.By(fmt.Sprintf("Wait for PD phase is %q", v1alpha1.UpgradePhase))
+					err = wait.PollImmediate(10*time.Second, 3*time.Minute, func() (bool, error) {
+						tc, err := cli.PingcapV1alpha1().TidbClusters(ns).Get(tcName, metav1.GetOptions{})
+						if err != nil {
+							log.Logf("failed to get TidbCluster %s/%s: %v", ns, tcName, err)
+							return false, nil
+						}
+						if tc.Status.PD.Phase != v1alpha1.UpgradePhase {
+							log.Logf("tc.Status.PD.Phase = %q, not %q yet", tc.Status.PD.Phase, v1alpha1.UpgradePhase)
+							return false, nil
+						}
+						return true, nil
+					})
+					framework.ExpectNoError(err, "failed to wait for PD phase")
+
+					ginkgo.By(fmt.Sprintf("Scale %s PD while in %q phase", op, v1alpha1.UpgradePhase))
+					err = controller.GuaranteedUpdate(genericCli, tc, func() error {
+						if op == "in" {
+							tc.Spec.PD.Replicas = 3
+						} else {
+							tc.Spec.PD.Replicas = 5
+						}
+						return nil
+					})
+					framework.ExpectNoError(err, "failed to scale %s PD", op)
+					err = oa.WaitForTidbClusterReady(tc, 10*time.Minute, 10*time.Second)
+					framework.ExpectNoError(err, "failed to wait for TidbCluster ready: %s/%s", ns, tc.Name)
+
+					ginkgo.By("Check PD replicas")
+					tc, err = cli.PingcapV1alpha1().TidbClusters(ns).Get(tcName, metav1.GetOptions{})
+					framework.ExpectNoError(err, "failed to get TidbCluster %s/%s: %v", ns, tcName, err)
+					if op == "in" {
+						framework.ExpectEqual(int(tc.Spec.PD.Replicas), 3)
+					} else {
+						framework.ExpectEqual(int(tc.Spec.PD.Replicas), 5)
+					}
+				})
+			}
+		})
+
+		// similar to PD scale-in/scale-out case above, need to check no evict leader scheduler left
```

Comment:
The evict leader scheduler is not checked.

Reply:
fixed in next commits.